scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/package.scala

```diff
@@ -2,18 +2,102 @@ package scala.meta
 package internal
 
 import java.nio.charset.Charset
-import scala.util.Try
+import java.nio.file.Files
+import scala.collection.immutable.Seq
 import scala.meta.inputs.{Input => mInput}
 import scala.meta.inputs.{Position => mPosition}
+import scala.meta.internal.io.FileIO
 import scala.meta.internal.io.PathIO
+import scala.meta.internal.semantic.vfs.Paths
 import scala.meta.internal.semantic.{schema => s}
 import scala.meta.internal.semantic.{vfs => v}
 import scala.meta.io._
 import scala.meta.{Dialect => mDialect}
 import scala.meta.{semantic => m}
 
 package object semantic {
-  implicit class XtensionDatabaseSchema(mdatabase: m.Database) {
+
+  implicit class XtensionSchemaDatabase(sdatabase: s.Database) {
+    private def entries = sdatabase.entries
+    def save(targetroot: AbsolutePath): Unit = {
+      sdatabase.entries.foreach { entry =>
+        val relpath = v.Paths.scalaToSemanticdb(RelativePath(entry.filename))
+        val abspath = targetroot.resolve(relpath).toNIO
+        Files.createDirectories(abspath.getParent)
+        val fos = Files.newOutputStream(abspath)
+        try {
+          s.Database(List(entry)).writeTo(fos)
+        } finally {
+          fos.close()
+        }
+      }
+    }
```

Comment:
With the previous vfs design, we had `Array[Byte]` + `s.Database` + `m.Database` loaded into memory for all .semanticdb files at some point. This is prohibitively expensive when running large-scale analysis. We want to avoid materializing an `Array[Byte]` entirely. Protos can be loaded in a streaming fashion, tag by tag, and that's what this refactoring did.